### PR TITLE
src: A RESTing change

### PR DESCRIPTION
**Pull request by vladfrangu**
**Please describe the changes this PR makes and why it should be merged:**

This PR touches REST in funny ways and fixes some bugs that could make you lock up a handler, implements bucket support, gives everyone more useful error stack traces, yada yada, let me submit this so I can get testers and write a comprehensive changelog

## Vlad, you're a mad man... What did you change now..

- Moved the API router from a getter in the REST Manager to directly in the client
	- 1 less getter :megu:
- Cleaned up APIRequest just a bit, and enforced `X-RateLimit-Precision` to `millisecond`
- Cleaned up the APIRouter proxy, and added the stack trace so we get a more useful error stack
- Implemented a Bucket with a lock since routes can now share the same Bucket
	- Thx Discord 
- Made HTTPError and DiscordAPIError use said stack trace
- Made the auth token be a static value
- Added a collection for buckets
~~- Separated each handler by its method as well~~
	~~- Except for reactions, which, due to us not knowing the bucket till we do a request, could give you a 429, so we just normalize it under a generic `*:/channels/123/messages/:id/reactions/*`~~
- Added some fixes to the RequestHandlers to make them use the Bucket locks
- Made the Client#rateLimit event emit some info about the bucket
- Added a few debug logs for certain things
- Enforced restTimeOffset to be a positive number, in case you were such a mad lad and made it negative, basically giving yourself 429's like they're cookies
- Added GuildEmojiStore#fetch because that was missing AND I needed it for the REST test suite

## What next?

I really need testers. I really mean it. This change affects the main thing that CAN AND WILL get you banned from Discord if you keep messing up. I need to test everything, every API call, every method, everything that is valid, to ensure:

1. No 429's encountered EVEN with restTimeOffset set to 0
2. No lockups of handlers
3. No unexpected behaviors
4. Proper error stack traces plz

### Ok, but how do _we_ test it?

Well, simple simple! Install the branch (`npm i vladfrangu/discOwOd.js#rest-rewrite`), turn on debug logs, and run your code... It really is that simple

If you encounter any 429s, please comment with a reproducible sample, and the debug logs!

**Status**
- [ ] Code changes have been tested against the Discord API, or there are no code changes
- [ ] I know how to update typings and have done so, or typings don't need updating

**Semantic versioning classification:**  
- [x] This PR changes the library's interface (methods or parameters added)
  - [x] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

**Comment by appellation** [comment]
We attempted to do async stack traces in #2744 which had to be reverted in #2848, so I am a bit skeptical of round 2.

**Comment by vladfrangu** [comment]
Fun fact, I tried implementing them like they were before in the PR you linked and... That caused the error name and message to vanish, and that unacceptable. If @Lewdcario can reproduce the same issue with this PR, I will remove them, but otherwise, I think they'll be fine. I understand your skepticism, and as this PR is still heavily WIP, things can and probably WILL change!

**Comment by didinele** [comment]
Should probably add `1 less var, now sitting at 0` 🎉

File: src/rest/APIRequest.js [hunk 1]
```diff
@@ -52,8 +52,8 @@ class APIRequest {
       headers['Content-Type'] = 'application/json';
     }
 
-    const timeout = this.client.setTimeout(() => controller.abort(), this.client.options.restRequestTimeout);
     const controller = new AbortController();
+    const timeout = this.client.setTimeout(() => controller.abort(), this.client.options.restRequestTimeout);
```
Comment (leRisen) [inline src/rest/APIRequest.js]: use `client` from restructuring `this.restManager`

```js
const timeout = client.setTimeout(() => controller.abort(), client.options.restRequestTimeout);
```

**Comment by NightYoshi370** [comment]
Is it me or do I get more ratelimits when I use this PR?

**Comment by vladfrangu** [comment]
> Is it me or do I get more ratelimits when I use this PR?

@NightYoshi370 anything is possible truthfully as this PR is still heavily WIP. Are you getting more 429's in debug logs? If so, what paths are getting 429s? Please put them on something like [Hastebin](https://hasteb.in)

**Comment by leRisen** [comment]
@vladfrangu hi! I use this PR for a very long time, it really fixes a lot of problems with API restriction, but there is one thing, the endpoint for add a reaction for this day returns an error code 429, and also in terms of logging messages, it’s also not ok, I occasionally get to 429 the code. Also, can you say why sometimes the role is not removed?

p.s Happy New Year xd

**Comment by vladfrangu** [comment]
@leRisen Hi! There's a very, very high chance that, by implementing the X-RateLimit-Bucket header, other things broke. It's..honestly a stupid header whose intention is good, but execution...leaves to desire. As such, in some commits coming soonish, I will revert that addition. If you want a more direct ping about that commit, find me in the Discord server (Vladdy#0002)

**Comment by wasdennnoch** [comment]
> It's..honestly a stupid header whose intention is good, but execution...leaves to desire.

Just curious, why? I haven't dug deep into the pitfalls yet.

**Comment by vladfrangu** [comment]
@wasdennnoch the hash isn't built with the major parameter. It's a generic hash for the route, not specific to a channel id, guild id, or webhook id. Thus, it's really not helpful when grouping routes with the same ratelimit bucket. I might bring this up to them later

**Comment by leRisen** [comment]
@vladfrangu little mistake, look please.

```
0|Kofuku  | TypeError: Cannot read property 'remaining' of undefined
0|Kofuku  |     at checkBucketUsable (C:\Users\Administrator\Desktop\Kofuku\node_modules\discord.js\src\rest\RequestHandler.js:25:17)
0|Kofuku  |     at RequestHandler.limited (C:\Users\Administrator\Desktop\Kofuku\node_modules\discord.js\src\rest\RequestHandler.js:49:16)
0|Kofuku  |     at RequestHandler.get isInactive [as isInactive] (C:\Users\Administrator\Desktop\Kofuku\node_modules\discord.js\src\rest\RequestHandler.js:44:47)
0|Kofuku  |     at handlers.sweep.handler (C:\Users\Administrator\Desktop\Kofuku\node_modules\discord.js\src\rest\RESTManager.js:27:48)
0|Kofuku  |     at Map.sweep (C:\Users\Administrator\Desktop\Kofuku\node_modules\discord.js\src\util\Collection.js:209:11)
0|Kofuku  |     at Timeout.RESTManager.client.setInterval [as _onTimeout] (C:\Users\Administrator\Desktop\Kofuku\node_modules\discord.js\src\rest\RESTManager.js:27:23)
0|Kofuku  |     at ontimeout (timers.js:436:11)
0|Kofuku  |     at tryOnTimeout (timers.js:300:5)
0|Kofuku  |     at listOnTimeout (timers.js:263:5)
0|Kofuku  |     at Timer.processTimers (timers.js:223:10)
```

**Comment by JMTK** [comment]
Would this PR solve any issues with "User aborted a request" 500 errors from causing the bot to eventually get rate limited? 

Right now I'm having to basically create my own bucket of errors and monitoring the number per minute because the bot isn't properly disconnecting from the gateway and continues to spam requests even though they return 500 "User aborted a request". 

> Added some fixes to the RequestHandlers to make them use the Bucket locks
> Made the Client#rateLimit event emit some info about the bucket

These 2 in particular seem like they would at least help me solve some issues or track down what's going on.

**Comment by vladfrangu** [comment]
This PR has grown very, very stale (and outdated). I am closing it for now, as REST improvements are coming to the `-next` repository 👀 . You may still use the branch at your own risk, I am **not** responsible for any issues caused by this in its current state. If you are still using this branch, I recommend moving off of it 👍

**Comment by Walkoud** [comment]
Package invalid, removed ?

**Comment by ImRodry** [comment]
Of course it's invalid, this PR is a year and a half old! You're probably looking for https://npmjs.com/pacjage/@discordjs/rest

**Pull request closed without merging**